### PR TITLE
feat(optics): throw better errors instead of simply undefined

### LINKED ISSUE #12: Better error messages on wrong usages

### PR DESCRIPTION
**Pull request by serras**
Fixes #12

### CODE REVIEW

File: src/Optional.js [hunk 1]
```diff
@@ -53,13 +57,19 @@ export const optional = curry((preview, set) => new Optional(preview, set))
...
+      if (key in obj) return obj[key]
+      else return notFound
+    },
     (val, obj) => (obj[key] ? { ...obj, [key]: val } : obj),
   )
 
 // optionalIx : Number → Optional s a
 export const optionalIx = (index) =>
   optional(
-    (obj) => obj[index] ?? notFound,
```
Comment (kutyel) [inline src/Optional.js:63]: Why this change? It looked better before 😢

File: src/operations.js [hunk 2]
```diff
@@ -111,25 +112,81 @@ export const optic = composeOptics
 export const path = composeOptics
 
 // reduce : Fold s a → (r -> a -> r) -> r -> s -> r
-export const reduce = curry((optic, f, i, obj) => optic.asFold.reduce(f, i, obj))
+export const reduce = curry((optic, f, i, obj) => {
+  if ('asFold' in optic) return optic.asFold.reduce(f, i, obj)
+  else
+    throw new UnavailableOpticOperationError(
```
Comment (kutyel) [inline src/operations.js:118]: This looks interesting! Thanks! 🎉 
Can you please add a unit test per exception with the matcher `.shouldThrow(\Exception\)`?
This way our test will be a proof that badly constructed lenses should the appropiate error etc 😉

**Review by kutyel — CHANGES_REQUESTED** [review]
Almost perfect! Only needs a few more tests with the new Exceptions! 😉

File: src/Optional.js (hunk 1 above)
Comment (serras) [reply src/Optional.js:63]: It turns out that some of the tools I use don't understand `??` yet. While doing so I also discovered a bug, because this won't work if `obj` was undefined.

File: src/operations.js (hunk 2 above)
Comment (serras) [reply src/operations.js:118]: You have them at the end of `operations.test.js`.

**Review by kutyel — DISMISSED** [review]
Great, thanks! can you also run?
```sh
yarn remove @babel/plugin-proposal-class-properties
```
This way that dep will be removed both from the `package.json` and the `yarn.lock` files! 😉

**Comment by serras** [comment]
I guess you mean the `nullish` plug-in, right? I did it, but nothing changes, maybe because when I added it I didn't add the package?
